### ISSUE TITLE
Implement S0i3 low power state

### ISSUE THREAD

**Issue by alice-sowerby** [FreeBSDFoundation/proj-laptop#32]
## User Story

As a laptop user I want to be able to put my laptop to sleep and for it to enter a low power state so that I can conserve battery charge and use the laptop for longer between charges.

NOTE: S0ix represents a set of SoC states which allow the SoC to consume less power by shutting off parts of the SoC when not needed.  [Source](https://www.intel.com/content/www/us/en/docs/socwatch/user-guide/2020/s0ix-states.html)

## Implementation Notes
There's an existing non-functional start of S0ix support, but nothing that is a demoable proof of concept.

## Acceptance Criteria

- [x] Can suspend/resume on AMD and enter a low power idle state, even if the CPU may sporadically wake.
- [ ] Test on Intel (may work out of the box)

## 15.0 Release Checklist

- [ ] Committed to the main branch and connected to the build (or for some tasks this means committed to the relevant repo to be delivered via ports/packages). 
- [ ] Appropriate test coverage or testing (automated or manual)
- [ ] Release notes (add directly to the file in the repository)
- [ ] Documentation in handbook and/or other appropriate places.

## Discussion thread
https://lists.freebsd.org/archives/freebsd-desktop/2025-March/005616.html

**Comment by obiwac** [comment]
Current state:

Driver for the SPMC (system power management controller, `acpi_spmc`) has been written ([D48387](https://reviews.freebsd.org/D48387)), and is able to parse and verify device power constraints as well as provide hooks for S0ix suspend entry/exit. These hooks will be used before and after suspending to idle (see #35) when the user wants to go to sleep.

D3cold support ([https://reviews.freebsd.org/D48384](D48384)) was added and fixes for switching device power states ([https://reviews.freebsd.org/D48385](D48385)) were made to respect the device constraints necessary for entering lower power idle states.

Currently, the only blocking thing is in ACPICA upstream (see https://github.com/acpica/acpica/pull/993), as D3cold and D3hot are not currently being distinguished and it is unclear whether the existing D3 should really be D3hot or D3cold.

This has been tested and is working on the AMD Framework laptop.

**Comment by alice-sowerby** [comment]
Looks like https://reviews.freebsd.org/D48385 is still sitting in review since January 2025. @emaste raised it in the Source Management team meeting but it still requires attention. 

This could be merged after testing and receive post-commit review.

**Comment by alice-sowerby** [comment]
@obiwac there are some comments on this patch from Warner. Let us know if you need any more input/support.

**Comment by obiwac** [comment]
> @obiwac there are some comments on this patch from Warner. Let us know if you need any more input/support.

Yes, I still need to revisit these - currently I have been mostly focused on the USB4 patches.

**Comment by obiwac** [comment]
No particular updates on these issues for April.

**Comment by alice-sowerby** [comment]
Note to readers: Take a look at the subissues which have updates.

**Comment by alice-sowerby** [comment]
@obiwac we took a look at the upstream open PR and noted that you are getting regular updates from them. Let us know if you would like us to help in any way.

**Comment by obiwac** [comment]
@alice-sowerby Yup, I am waiting for a verdict on their end :)

**Comment by alice-sowerby** [comment]
@obiwac the upstream PR looks like it could do with another ping, and maybe mention to them about the release schedule we are trying to hit?

**Comment by obiwac** [comment]
S0i3 works and am now putting CPUs in C3. There is currently an issue with CPU0 sending IPIs to another CPU to handle the GPE it received, which wakes it. This happens because we add the GPE handler task to a taskqueue which gets executed asynchronously. I'm looking at either executing this synchronously or at least binding it to CPU0 and somehow waiting for the GPE task to conclude before idling CPU0 again.

**Comment by dumbbell** [comment]
We discussed by email with @obiwac. He was able to figure out the problem before I could come back to FreeBSD work today.

**Comment by obiwac** [comment]
See here for work done on S0i3 during December: https://github.com/FreeBSDFoundation/status-updates/blob/main/Aymeric_Wibo/2025-12.md

**Comment by alice-sowerby** [comment]
Removed the 15.1 milestone as it will not be possible to complete it in time.

**Comment by obiwac** [comment]
Most time spent on this was debugging the NVMe issue - still no dice unfortunately, though I think our D-state code (again) needs a deeper look at.

olce@ has been spending some time looking at the `acpi_spmc` driver and generally improving it (see the stack at [D56798](https://reviews.freebsd.org/D56798)), and thj@ has been trying to get things working on his laptop (see [D56619](https://reviews.freebsd.org/D56619), which adds Krackan Point support to the `amdsmu` driver).

**Comment by OlCe2** [comment]
In-flight improvements to the `acpi_spmc(4)` driver are in the stack surrounding [D56756](https://reviews.freebsd.org/D56756) (this is the same stack as mentioned by obiwac@ above).

Among these changes, we highlight these:
- [acpi_spmc(4): Human-readably print supported DSMs and their functions](https://reviews.freebsd.org/D56756)
- [acpi_spmc(4): Gracefully support only the Microsoft DSM](https://reviews.freebsd.org/D56813)
- [acpi_spmc(4): Enable multiple instances](https://reviews.freebsd.org/D56817)
- [acpi_spmc(4): Add a sysctl knob to request verbosity](https://reviews.freebsd.org/D56875)
- [acpi_spmc(4): Be less verbose by default](https://reviews.freebsd.org/D56876)
- [acpi_spmc(4): Only run DSM functions reported present](https://reviews.freebsd.org/D56879)
- [acpi_spmc(4): Auto-detect DSM revisions by default](https://reviews.freebsd.org/D56882)

**Comment by OlCe2** [comment]
The stack mentioned in the previous comment has just been committed.  There's still a few things to do in `acpi_spmc(4)` in order to harden it in case the firmware does not obey some spec (Intel) or already existing tables (AMD) to report constraints.

**Comment by obiwac** [comment]
Still struggling through getting our D-state code to be a little more sane and hopefully to fix our NVMe suspend issue, but in the meantime mostly worked on userspace changes, notably renaming the sleep types ([D56920](https://reviews.freebsd.org/D56920)), adding a new power management interface under `/dev/power` ([D55508](https://reviews.freebsd.org/D55508)), and rewriting zzz(8) to use this new interface ([D56918](https://reviews.freebsd.org/D56918)).

@OlCe2 has also committed a bunch of fixes and quality-of-life improvements to the `acpi_spmc` driver as well as looking into an Intel PMC driver (equivalent to amdsmu), and @adventureloop has done some testing and debugging on his laptop.

A revision for an amdsmu(4) manual page is up ([D57366](https://reviews.freebsd.org/D57366)), and I am working on one for `acpi_spmc`, and I generally want to explain S0ix and suspend-to-idle a little better.

**Comment by obiwac** [comment]
A little more D-state debugging, but not much else for the month of June.

**Comment by obiwac** [comment]
Confirmed that the D-states should be set entirely correctly for the NVMe, so the issue is elsewhere. I have switched tacks to trying to emulate the same kind of issue on a minimal Linux kernel, similar to what I'd done for getting USB4 suspend to work.

**Comment by obiwac** [comment]
Am now able to resume NVMe drives 🎉 

The issue was indeed D-states - we were doing this in two places, and one of the places was setting it back to D3hot before suspend when it should've been in D3cold. I still need to clean this up and create a revision out of this.

Next up will be figuring out how to replace the hacky scheduler revision with something more permanent.
